tests/unittests/test_upgrade.py

```diff
@@ -39,7 +144,67 @@ def previous_obj_pkl(self, request):
         Test implementations _must not_ modify the ``previous_obj_pkl`` which
         they are passed, as that will affect tests that run after them.
         """
-        return pkl_load(str(request.param))
+        return sources.pkl_load(str(request.param))
+
+    @mock.patch.object(
+        importer,
+        "match_case_insensitive_module_name",
+        lambda name: f"DataSource{name}",
+    )
+    def test_all_ds_init_vs_unpickle_attributes(self, mocker, paths, tmpdir):
+        """Unpickle resets any instance attributes created in __init__
+
+        This test asserts that deserialization of a datasource cache
+        does proper initialization of any 'new' instance attributes
+        created as a side-effect of the __init__ method.
+
+        Without proper _unpickle coverage for newly introduced attributes,
+        the new deserialized instance will hit AttributeErrors at runtime.
```

Comment:
This is a great!

nit: This explains what the test validates, but not how. Usually that's not a problem, but since this is innovative and not typical of cloud-init test code, a sentence describing how it asserts the proper initialization would go along ways. A comment or two inline adding context wouldn't hurt either.